src/fcntl.rs

```diff
@@ -596,39 +599,120 @@ pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
     Errno::result(res)
 }
 
-// TODO: convert to libc_enum
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
 #[non_exhaustive]
 pub enum FlockArg {
     LockShared,
     LockExclusive,
-    Unlock,
     LockSharedNonblock,
     LockExclusiveNonblock,
-    UnlockNonblock,
 }
 
+/// Represents valid types for flock.
+///
+/// # Safety
+/// Types implementing this must be `!Clone`.
 #[cfg(not(any(target_os = "redox", target_os = "solaris")))]
-pub fn flock(fd: RawFd, arg: FlockArg) -> Result<()> {
-    use self::FlockArg::*;
+pub unsafe trait Flockable: AsRawFd {}
 
-    let res = unsafe {
-        match arg {
-            LockShared => libc::flock(fd, libc::LOCK_SH),
-            LockExclusive => libc::flock(fd, libc::LOCK_EX),
-            Unlock => libc::flock(fd, libc::LOCK_UN),
-            LockSharedNonblock => {
-                libc::flock(fd, libc::LOCK_SH | libc::LOCK_NB)
-            }
-            LockExclusiveNonblock => {
-                libc::flock(fd, libc::LOCK_EX | libc::LOCK_NB)
-            }
-            UnlockNonblock => libc::flock(fd, libc::LOCK_UN | libc::LOCK_NB),
+/// Represents an owned flock, which unlocks on drop.
+///
+/// See flock(2) for details on locking semantics.
+// `ManuallyDrop` is necessary to circumvent move out of `Drop` type error.
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+#[derive(Debug)]
+pub struct Flock<T: Flockable>(ManuallyDrop<T>);
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Drop for Flock<T> {
+    fn drop(&mut self) {
+        // Result is ignored because flock has no documented failure cases.
+        _ = unsafe { libc::flock(self.0.as_raw_fd(), libc::LOCK_UN) };
```

Comment:
We should call `ManuallyDrop::drop(&mut self.0)` here or the inner `T` will never be dropped

Reply:
Well, we should not call it or the following code will be a double drop:

```rs
let file = File;
let lock = Flock::lock(file, arg);
let file = lock.unlock() // file dropped within `lock.unlock()`


// file dropped here
```

Reply:
I kinda of think `ManuallyDrop` is a bad choice here, what about using `Flock<Option<T>>` here, this should solve the issue

Reply:
Concur. See latest changes.